File: Leetcode Daily Challenge/025-July-2025/019-remove-sub-folders-from-the-filesystem.cpp
// https://leetcode.com/problems/remove-sub-folders-from-the-filesystem/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    // TC : O(N*log(N))
    // SC : O(N)
    vector<string> approach_1(vector<string>& folder) {
        sort(begin(folder), end(folder));
        vector<string> result;
        unordered_set<string> st;
        for (string &fold : folder) {
            string prefix;
            bool isSubfolder = 0;
            for (char &ch : fold) {
                if (ch == '/' && st.count(prefix)) {
                    isSubfolder = 1;
                    break;
                }
                prefix += ch;
            }
            if (!isSubfolder) {
                st.insert(fold);
                result.push_back(fold);
            }
        }
        return result;
    }

    // TC : O(N*log(N))
    // SC : O(1)
    vector<string> approach_2(vector<string>& folder) {
        sort(begin(folder), end(folder));
        vector<string> result;
        for (string &fold : folder) {
            if (result.empty() || fold.find(result.back()) != 0 || fold[result.back().size()] != '/') {
                result.push_back(fold);
            }
        }
        return result;
    }
public:
    vector<string> removeSubfolders(vector<string>& folder) {
        // return approach_1(folder);
        return approach_2(folder);   // No Extra Space
    }
};

File: Leetcode Daily Challenge/025-July-2025/019-remove-sub-folders-from-the-filesystem.java
// https://leetcode.com/problems/remove-sub-folders-from-the-filesystem/




import java.util.*;

class Solution {
    public List<String> removeSubfolders(String[] folder) {
        // return approach_1(folder);
        return approach_2(folder);  // Extra Space Optimized
    }

    // TC : O(N*log(N))
    // SC : O(1)
    private List<String> approach_2(String[] folder) {
        Arrays.sort(folder);
        List<String> result = new ArrayList<>();
        for (String fold : folder) {
            if (result.isEmpty() || !fold.startsWith(result.get(result.size() - 1) + "/")) {
                result.add(fold);
            }
        }
        return result;
    }

    // TC : O(N*log(N))
    // SC : O(N)
    private List<String> approach_1(String[] folder) {
        Arrays.sort(folder);
        Set<String> st = new HashSet<>();
        List<String> result = new ArrayList<>();
        for (String fold : folder) {
            boolean isSubFolder = false;
            StringBuilder prefix = new StringBuilder();
            for (char ch : fold.toCharArray()) {
                if (ch == '/' && st.contains(prefix.toString())) {
                    isSubFolder = true;
                    break;
                }
                prefix.append(ch);
            }
            if (!isSubFolder) {
                st.add(prefix.toString());
                result.add(fold);
            }
        }
        return result;
    }
}